### PR TITLE
Fix top-level activation for Move Static Members

### LINKED ISSUE #59470: Move static members to another type is suggested for top-level statements
**Version Used**:  
VS 17.1.0 Preview 6.0, '4.1.0-3.22075.3 (592501cb)'. Language version: 10.0.

**Steps to Reproduce**:

1. dotnet new console
2. Write a statement such as `_ = 42;`
3. Open the quick actions menu

image

image

````
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp.PullMembersUpOptionsBuilder.<>c__DisplayClass0_0.<BuildPullMembersUpOptions>b__0(ValueTuple`2 memberAndMakeAbstract)
   at System.Collections.Immutable.ImmutableArray.CreateRange[TSource,TResult](ImmutableArray`1 items,Func`2 selector)
   at Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp.PullMembersUpOptionsBuilder.BuildPullMembersUpOptions(INamedTypeSymbol destination,ImmutableArray`1 members)
   at async Microsoft.CodeAnalysis.MoveStaticMembers.MoveStaticMembersWithDialogCodeAction.ComputeOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.CodeActions.CodeActionWithOptions.GetOperationsAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.InvokeWorkerAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformActionAsync(<Unknown Parameters>)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
````

**Expected Behavior**:

Move static members to another type is not suggested

**Actual Behavior**:

Move static members to another type is suggested, but fails

### PR DESCRIPTION
**Pull request by Soreloser2**
Fixes #59470

Change selection logic for Move Static Members to follow more closely with Extract Class and Pull Members Up, where we get the enclosing type and selected symbol from the selected syntax node instead of the selection span and its members. This fixes the problem where the refactoring was being suggested on top-level statements.

Offers a slight change in the dialog box behavior, in that it will now preselect the member that triggered the activation, instead of the first member of the enclosing class.

Added tests to document non-activation in top-level statement.

### CODE REVIEW

**Comment by Soreloser2** [comment]
Currently a few tests around field selection are expected to fail until #62102 is merged, as that PR changes how the members are obtained in logic that is shared among multiple refactoring providers.

File: src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs [hunk 1]
```diff
@@ -38,26 +38,16 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                 return;
             }
 
-            var selectedType = semanticModel.GetEnclosingNamedType(span.Start, cancellationToken);
-            if (selectedType == null)
+            var selectedMember = semanticModel.GetDeclaredSymbol(memberDeclaration, cancellationToken);
+            if (selectedMember is null || selectedMember.ContainingType is null ||
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs:42]:
```suggestion
            if (selectedMember?.ContainingType is null ||
```

File: src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs [hunk 2]
```diff
@@ -38,26 +38,16 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
...
-            {
-                return;
-            }
-
-            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
-
-            var action = new MoveStaticMembersWithDialogCodeAction(document, span, service, selectedType, context.Options, selectedMember: selectedMembers[0]);
+            var action = new MoveStaticMembersWithDialogCodeAction(document, memberDeclaration.Span, service, selectedMember.ContainingType, context.Options, selectedMember: selectedMember);
 
-            context.RegisterRefactoring(action, selectedMembers[0].DeclaringSyntaxReferences[0].Span);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs:60]: why this change?

File: src/EditorFeatures/CSharpTest/MoveStaticMembers/CSharpMoveStaticMembersTests.cs [hunk 3]
```diff
@@ -2565,6 +2597,44 @@ public class Class1
...
+            var test = new Test("", ImmutableArray<string>.Empty)
+            {
+                TestState =
+                {
+                    OutputKind = OutputKind.ConsoleApplication
+                },
+                TestCode = initialMarkup,
+                FixedCode = initialMarkup,
+            };
+            test.LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/MoveStaticMembers/CSharpMoveStaticMembersTests.cs:2617]:
```suggestion
                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9;
            };
```
?

File: src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs (hunk 2 above)
Comment (Soreloser2) [reply src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs:60]: Ah, I incorrectly assumed that the overload without a provided span used the Span property on the provided `CodeAction`, but I see it fills in null if not provided. I'll change it to `memberDeclaration.Span`. I changed it initially because we want to make changes on the member we selected, not just the first reference to the first member of the type.

Comment (Soreloser2) [reply src/Features/Core/Portable/MoveStaticMembers/AbstractMoveStaticMembersRefactoringProvider.cs:60]: More generally, I removed that code above (where we get all the members and validate each one) because if we are currently selecting a valid member for refactoring, then we have at least one member that is valid, so we don't need to check all the members.

File: src/EditorFeatures/CSharpTest/MoveStaticMembers/CSharpMoveStaticMembersTests.cs [hunk 4]
```diff
@@ -2325,6 +2325,38 @@ internal static class Class1Helpers
             await TestMovementNewFileAsync(initialMarkup, expectedResult1, expectedResult2, newFileName, selectedMembers, selectedDestinationName).ConfigureAwait(false);
         }
 
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsMoveStaticMembers)]
+        public async Task TestSelectInKeyWordOfMulitFieldDeclaration()
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/MoveStaticMembers/CSharpMoveStaticMembersTests.cs:2329]:
```suggestion
        public async Task TestSelectInKeywordOfMulitFieldDeclaration()
```

Not a big deal, but we treat "keyword" as one world in Roslyn.

**Review by jasonmalinowski — APPROVED** [review]
